### DISCUSSION TITLE
Desktop ARM macOS Sonoma / openjdk "21.0.1" 2023-10-17: crash on start

### DISCUSSION THREAD

**Discussion by Klaus-Tockloth** [devemux86/cruiser#353] [Q&A] [answered]
Cruiser doesn't start with 'openjdk "21.0.1" 2023-10-17' under 'macOS Sonoma' (14.1, Apple-Silicon).

Could someone confirm this behavior?

Tested with:

- cruiser-3.0.23
- cruiser-3.0.32

Logfile: [hs_err_pid4885.log](https://github.com/devemux86/cruiser/files/13185950/hs_err_pid4885.log)

**Comment by devemux86** [comment]
@Klaus-Tockloth thanks for the report.

Is it about Mapsforge or VTM (OpenGL)?
VTM (OpenGL) currently does not work on Apple ARM until someone with Mac ARM builds the native libs:
- https://github.com/mapsforge/vtm/issues/1004

Can try another Java, like [OpenJDK 17 (Temurin)](https://adoptium.net/).

**Comment by Klaus-Tockloth** [comment]
I'm still using Mapsforge (and have currently no plans to switch to VTM).

A downgrade to 'openjdk version "17.0.9" 2023-10-17' doesn't help:

```
% sh cruiser.sh 
2023-10-27 10:06:07.890 java[5545:124699] WARNING: Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application. Opt-in to secure coding explicitly by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState:.
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00000001031b2d58, pid=5545, tid=86275
#
# JRE version: OpenJDK Runtime Environment Homebrew (17.0.9) (build 17.0.9+0)
# Java VM: OpenJDK 64-Bit Server VM Homebrew (17.0.9+0, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [libawt.dylib+0x2ed58]  fillAAPgram+0x3e0
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/klaustockloth/Hobby/fzk-android/cruiser-3.0.32/hs_err_pid5545.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/Homebrew/homebrew-core/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
cruiser.sh: line 3:  5545 Abort trap: 6           java -Xmx6024M -jar cruiser.jar
```

Logfile: [hs_err_pid5545.log](https://github.com/devemux86/cruiser/files/13186532/hs_err_pid5545.log)

**Comment by devemux86** [comment]
You can see here my sad conclusion about the future of Mapsforge **Android engine**:
- https://github.com/mapsforge/mapsforge/issues/1437#issuecomment-1683993596

---

About the desktop, you could try Mapsforge desktop samples and see if it can work on Mac ARM.
(see instructions [here](https://github.com/mapsforge/mapsforge/blob/master/docs/Desktop.md))

I work on Linux and fortunately everything works here. :slightly_smiling_face:

**Comment by Klaus-Tockloth** [comment]
A Google search for "WARNING: Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application." finds a lot of relations to macOS 14 (Sonoma).

BTW: Before macOS 14 I'm using 'Cruiser Desktop' for a long time on my M1-Mac without any problem. The defect (crash on startup) came probably in with the update to 'Sonoma'.

**Comment by devemux86** [comment]
Have you tried with an external Java, like the [OpenJDK 17 (Temurin)](https://adoptium.net/)?

**Comment by Klaus-Tockloth** [comment]
Same result (crash on startup) with OpenJDK 21 and OpenJDK 17:

```
% sh cruiser.sh 
2023-10-27 13:39:22.494 java[7852:198017] WARNING: Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application. Opt-in to secure coding explicitly by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState:.
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x000000010582efc8, pid=7852, tid=53511
#
# JRE version: OpenJDK Runtime Environment Homebrew (21.0.1) (build 21.0.1)
# Java VM: OpenJDK 64-Bit Server VM Homebrew (21.0.1, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [libawt.dylib+0x2efc8]  fillAAPgram+0x3e0
...
```

```
% sh cruiser.sh                             
2023-10-27 13:40:10.362 java[7862:198532] WARNING: Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application. Opt-in to secure coding explicitly by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState:.
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x000000010582ed58, pid=7862, tid=85763
#
# JRE version: OpenJDK Runtime Environment Homebrew (17.0.9) (build 17.0.9+0)
# Java VM: OpenJDK 64-Bit Server VM Homebrew (17.0.9+0, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [libawt.dylib+0x2ed58]  fillAAPgram+0x3e0
```

**Reply by devemux86** [reply]
The "Temurin" does not appear in the Java info. It seems you still run Cruiser with internal Java.
It is not enough to install the external Java. Must also use its `java` command to run the programs.
Edit the cruiser.sh file and change the `java` path or run Cruiser from the command line.

For example on Linux, the `java -version` returns:
```
openjdk version "17.0.8" 2023-07-18
OpenJDK Runtime Environment Temurin-17.0.8+7 (build 17.0.8+7)
OpenJDK 64-Bit Server VM Temurin-17.0.8+7 (build 17.0.8+7, mixed mode, sharing)
```

**Comment by Klaus-Tockloth** [comment]
Hmm, 'java -version' shows this:

```
% java -version
openjdk version "17.0.9" 2023-10-17
OpenJDK Runtime Environment Homebrew (build 17.0.9+0)
OpenJDK 64-Bit Server VM Homebrew (build 17.0.9+0, mixed mode, sharing)
```
```
% java -version
openjdk version "21.0.1" 2023-10-17
OpenJDK Runtime Environment Homebrew (build 21.0.1)
OpenJDK 64-Bit Server VM Homebrew (build 21.0.1, mixed mode, sharing)
```

**Reply by devemux86** [reply]
Use the `java` with full path from the folder where the OpenJDK (Temurin) is installed.

**Comment by Klaus-Tockloth** [comment]
BTW: 'Homebrew' is a package manager for macOS. It seems that the JDK is build by 'Homebrew'. In other words: I have no 'Temurin OpenJDK'.

```
% /opt/homebrew/opt/openjdk@17/bin/java -version
openjdk version "17.0.9" 2023-10-17
OpenJDK Runtime Environment Homebrew (build 17.0.9+0)
OpenJDK 64-Bit Server VM Homebrew (build 17.0.9+0, mixed mode, sharing)
```

```
% /opt/homebrew/opt/openjdk/bin/java -version
openjdk version "21.0.1" 2023-10-17
OpenJDK Runtime Environment Homebrew (build 21.0.1)
OpenJDK 64-Bit Server VM Homebrew (build 21.0.1, mixed mode, sharing)
```

**Reply by devemux86** [reply]
OpenJDK (Temurin) allows to download archives ".tar.gz" with the Java.
Could extract them and run the Java from inside their folders.
At least this works on Windows / Linux without an installation.

https://adoptium.net/temurin/releases/

**Comment by Klaus-Tockloth** [comment, marked as answer]
After the installation of Oracle-JDK, 'Cruiser (3.0.32) desktop for macOS' works again:

/Library/Java/JavaVirtualMachines/jdk-21.jdk/Contents/Home/bin/java -jar cruiser.jar

<img width="908" alt="Bildschirmfoto 2023-10-28 um 08 48 53" src="https://github.com/devemux86/cruiser/assets/5493128/e53b8996-8642-4f21-a85f-89036cee52c3">

```
% /Library/Java/JavaVirtualMachines/jdk-21.jdk/Contents/Home/bin/java -version 
java version "21.0.1" 2023-10-17 LTS
Java(TM) SE Runtime Environment (build 21.0.1+12-LTS-29)
Java HotSpot(TM) 64-Bit Server VM (build 21.0.1+12-LTS-29, mixed mode, sharing)
```

**Reply by devemux86** [reply]
@Klaus-Tockloth indeed that's how you need to run an external java, using a full path.
So it seems that companies also modify how Java works in their systems (like Android).

Interesting that Oracle Java works on your device.
There were issues on Windows / Linux, that's why we use an external OpenJDK.

Have you also tried [OpenJDK Temurin](https://adoptium.net/temurin/releases/) (with full path)?
